### ISSUE TITLE
Cosine similarity in Figure 1

### ISSUE THREAD

**Issue by Lycus99** [ljwztc/CLIP-Driven-Universal-Model#25]
Nice work!

I am trying to repeat the cosine similarity you showed in Figure 1. However, I found that the cosine similarity I computed by the CLIP text encoder differs from yours.  The code about extracting the text embeddings and computing cosine similarity is listed.

`
import clip
import torch
from sklearn.metrics.pairwise import cosine_similarity

device = "cuda" if torch.cuda.is_available() else "cpu"
model, _ = clip.load("ViT-B/32", device=device)

text = ['Liver Tumor', 'Hepatic Vessel', 'Right Kidney', 'Left Kidney', 'Kidney Tumor', 'Liver']
input_text1 = ['A photo of a ' + i for i in text]
input_text2 = ['There is {} in this computerized tomography'.format(i) for i in text]
input_text3 = ['A computerized tomography of a ' + i for i in text]

with torch.no_grad():
    input_text1 = clip.tokenize(input_text1).to(device)
    features1 = model.encode_text(input_text1)
    np.save('universal_model_clip_v1.npy', features1.cpu().detach().numpy())

    input_text2 = clip.tokenize(input_text2).to(device)
    features2 = model.encode_text(input_text2)
    np.save('universal_model_clip_v2.npy', features2.cpu().detach().numpy())

    input_text3 = clip.tokenize(input_text3).to(device)
    features3 = model.encode_text(input_text3)
    np.save('universal_model_clip_v3.npy', features3.cpu().detach().numpy())
    print('1')

cls_weight = np.load('universal_model_clip_v3.npy')
sim3 = cosine_similarity(cls_weight)
plt.imshow(sim3, cmap='viridis', interpolation='nearest')
plt.colorbar()
plt.show()
`
The similarity is shown in the following figure.

![image](https://github.com/ljwztc/CLIP-Driven-Universal-Model/assets/109274751/09c451b4-f023-410f-85a0-5713c7a0886e)

For example, the similarity score between "Liver Tumor" and "Right Kidney" is close to the score between "Liver Tumor" and "Hepatic Vessel". However, as shown in Figure 1, their values are very different. What method did you use to compute the cosine similarity score? Thanks very much.

**Comment by Lycus99** [comment]
![image](https://github.com/ljwztc/CLIP-Driven-Universal-Model/assets/109274751/d29d0e20-81e0-4fe9-bd6d-173dc145a619)

Figure 1 in the paper.

**Comment by Lycus99** [comment]
Thanks for your reply. I used the following code to normalize the text embedding, but the result is still different from yours. My text embedding fails to model the relationship.

`def normalize_(array, axis):
    mean = np.mean(array, axis=axis, keepdims=True)
    std = np.std(array, axis=axis, keepdims=True)
    normalized_array = (array - mean) / std
    return normalized_array

cls_weight = np.load('universal_model_clip_v3.npy')
cls_weight = normalize_(cls_weight, 0)
sim = cosine_similarity(cls_weight)

cls_weight = np.load('universal_model_clip_v3.npy')
cls_weight = normalize_(cls_weight, 1)
sim = cosine_similarity(cls_weight)
`

**Comment by ljwztc** [comment]
hi. I remember that the original results were not obvious and the relationship become apparent after normalization. I have checked my codebase but didn't find the code. The experimental records are attached for your reference. 
<img width="1042" alt="image" src="https://github.com/ljwztc/CLIP-Driven-Universal-Model/assets/36372917/4acef090-9047-4c29-8cdd-96c56b420ee8">
